### PR TITLE
#162948991 Reset Password via email

### PR DESCRIPTION
**Pull request by Nicanor008**
#### What does this PR do?
Allows users to reset password using email

#### Description of Task to be completed?
- Confirms if user exists on provision of emails
- Generate token for the email provided
- Send an email with a password reset link to the email provided
- Allow user to reset new password and confirm that the password matches the provided

#### How should this be manually tested?
* After cloning the repo using `git clone https://github.com/andela/ah-fulldeck.git`
* Move into the project folder `cd ah-fulldecck`
* Check out this branch `git checkout ft-resetPassword-viaEmail-162948991`. 
* Add the following to the .env file
```
export EMAIL_SENDER="ahfulldeck@gmail.com"
export EMAIL_HOST="smtp.gmail.com"
export EMAIL_HOST_USER="ahfulldeck"
export EMAIL_HOST_PASSWORD="Ful!deck1234"
export EMAIL_PORT=587
```
* Run the tests using the command `python manage.py test`
* Run the server `python manage.py runserver`
* Register a user, open this endpoint `127.0.0.1:8000/api/v1/users/password_reset/` to allow the registered user to reset email.

#### Any background context you want to provide?
Email verification features is achieved by google SMTP server host.

#### What are the relevant pivotal tracker stories?
[#162948991](https://www.pivotaltracker.com/story/show/162948991)

#### Successful email verification and sending of password reset link
screen shot 2019-01-16 at 17 15 18

#### Email inbox

screenshot 2019-01-21 at 10 06 27

#### After clicking the link sent to you, this is the endpoint looks
screen shot 2019-01-16 at 17 20 07

### CODE REVIEW

**Comment by Hesbon5600** [comment]
In your PR, instead of 'Move into the project folder cd <path_of_the_project>', just say 'cd ah-fulldeck' because the project folder is 'ah-fulldeck'
Also, include a screenshot of the mail inbox that prompts the user to click in order to reset their password

**Comment by Nicanor008** [comment]
> In your PR, instead of 'Move into the project folder cd <path_of_the_project>', just say 'cd ah-fulldeck' because the project folder is 'ah-fulldeck'
> Also, include a screenshot of the mail inbox that prompts the user to click in order to reset their password

Done @Hesbon5600 . Thanks for the feedback

**Review by kzyangiro — CHANGES_REQUESTED** [review]
@Nicanor008 @eduhmik 
1. Fix tests, coverage decreased. 
2. Fix your commits to show only one, your work and only the files you changed

File: .travis.yml [hunk 1]
```diff
@@ -10,6 +10,9 @@ install:
 services:
  - postgresql
 
+
+services:
+ - postgresql
 # Run tests
```
Comment (kzyangiro) [inline .travis.yml]: @Nicanor008 A duplicate here

File: authors/apps/authentication/serializers.py [hunk 2]
```diff
@@ -148,6 +169,12 @@ class UserSerializer(serializers.ModelSerializer):
         write_only=True
     )
 
+    # confirm_password = serializers.CharField(
+    #     max_length=128,
+    #     min_length=8,
+    #     write_only=True
+    # )
+
     class Meta:
```
Comment (kzyangiro) [inline authors/apps/authentication/serializers.py]: @Nicanor008 Unnecessary comments here.

File: authors/apps/authentication/views.py [hunk 3]
```diff
@@ -26,7 +27,79 @@ def get(self, request):
...
+            }
+            return Response(result, status=status.HTTP_404_NOT_FOUND)
+
+
+class PasswordUpdateAPIView(generics.UpdateAPIView):
+    permission_classes = (AllowAny,)
+    serializer_class = PasswordResetSerializer
+    # The URL conf should include a keyword argument corresponding to this value
+    look_url_kwarg = 'token'
+
```
Comment (kzyangiro) [inline authors/apps/authentication/views.py:60]: @Nicanor008 This was to be changed.

File: authors/settings.py [hunk 4]
```diff
@@ -170,4 +179,4 @@
 EMAIL_HOST_USER = os.getenv('EMAIL_HOST_USER')
 EMAIL_HOST_PASSWORD = os.getenv('EMAIL_HOST_PASSWORD')
 EMAIL_PORT = os.getenv('EMAIL_PORT')
-EMAIL_USE_TLS = True
\ No newline at end of file
+EMAIL_USE_TLS = True
```
Comment (kzyangiro) [inline authors/settings.py]: @Nicanor008 @eduhmik Code duplication, not acceptable.

File: authors/apps/authentication/templates/reset_password.html [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        <p style="text-align: left;font-size: 16px;color: #808080;padding-top: 16px;">Hello!</p>
+        <p style="text-align: left;font-size: 16px;color: #808080;padding-top: 16px;">You are receiving this email because we received a password reset request for your account on <a href="https://ah-fulldeck-staging.herokuapp.com/" style="font-weight: 800;color: orange;text-decoration: none;">
+            Authors Haven.</a></p>
+        <p style="text-align: left;font-size: 16px;color: #808080;padding-top: 16px;">Please  <a href="{{reset_link}}" style="font-weight: 800;color: orange;text-decoration: none;">click Here</a> to reset password.</p>
+        <p style="text-align: left;font-size: 16px;color: #808080;padding-top: 16px;">Regards,</p>
+        <p style="text-align: left;font-size: 16px;color: #808080;padding-top: 16px;">Fulldeck Team &copy; 2019</p>
+        <h6 style="color: grey;font-family: Helvetica,Arial,sans-serif;font-size: 10px;font-weight: 400;line-height: 1.3;margin: 0;margin-bottom: 10px;padding: 0;text-align: right;padding-top: 16px;">&copy; 2019 Authors Haven .<br> All Rights Reserved</h6>
+      </div>
+
+  </body>
```
Comment (kzyangiro) [inline authors/apps/authentication/templates/reset_password.html]: @Nicanor008 Full duplication. Code not dry. Unacceptable.

File: authors/settings.py (hunk 4 above)
Comment (Nicanor008) [reply authors/settings.py]: Removed duplicates

File: authors/apps/authentication/templates/reset_password.html (hunk 5 above)
Comment (Nicanor008) [reply authors/apps/authentication/templates/reset_password.html]: Working on this to create a base file

File: .travis.yml [hunk 6]
```diff
@@ -10,6 +10,9 @@ install:
 services:
  - postgresql
 
+
```
Comment (Nyakaru) [inline .travis.yml]: Hey Nick you have a repetition of services

File: authors/apps/authentication/tests/base_test.py [hunk 7]
```diff
@@ -16,6 +19,14 @@ def setUp(self):
                 'password': 'TestUser123'
             }}
 
+            # User.is_active = True
+        email = 'test@user.com'
+        # user_data = User.objects.filter(email=email).first()
+        token = jwt.encode({"email": "test@user.com"},
```
Comment (kzyangiro) [inline authors/apps/authentication/tests/base_test.py:23]: @Nicanor008 Remove these two comments

File: authors/apps/authentication/tests/test_password_update.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+            b'Password must be at least 8 characters long', response.content)
+
+    def test_unique_password(self):
+        TestPasswordReset.test_successful_email_sent(self)
+        response = self.client.put(self.password_update_url,
+                                   data={"password": "Nicanor",
+                                         "confirm_password": "Nicanor"},
+                                   format='json')
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+        self.assertIn(
```
Comment (kzyangiro) [inline authors/apps/authentication/tests/test_password_update.py]: @Nicanor008 A duplicate in all 400 status code tests, write a single method to be used in all those tests

File: .travis.yml [hunk 9]
```diff
@@ -10,6 +10,9 @@ install:
...
+services:
+ - postgresql
```
Comment (kzyangiro) [inline .travis.yml]: @Nicanor008 Still a duplicate here.

Comment (Nicanor008) [reply .travis.yml]: Had forgotten, done

File: authors/apps/authentication/tests/test_password_update.py (hunk 8 above)
Comment (Nicanor008) [reply authors/apps/authentication/tests/test_password_update.py]: I have implemented this

File: authors/apps/authentication/tests/base_test.py (hunk 7 above)
Comment (Nicanor008) [reply authors/apps/authentication/tests/base_test.py:23]: I have removed the unnecessary comments

File: .travis.yml (hunk 6 above)
Comment (Nicanor008) [reply .travis.yml]: got that, thanks

**Review by kzyangiro — APPROVED** [review]

**Pull request merged**